### DISCUSSION TITLE
filter by ids and then find the topK by a given vector

### DISCUSSION THREAD

**Discussion by chenchuangc** [milvus-io/milvus#11622] [Q&A and General discussion]
sorry for bother you 
## I want use milvus help me do this work
1. i have a collection have about 6million vectors like this 
![image](https://user-images.githubusercontent.com/13183268/141223846-4fde1909-120f-4a6c-95f1-75b9d0067a57.png)
2. my query is like this 
```
        vectorParam := entity.FloatVector(some_test_vector)  // 64 dimensions  
	idFilter := "ID in [ 123,234,456 ]"  // the ids may be more than 5000
	// Use flat search param
	sp, _ := entity.NewIndexFlatSearchParam(10)
	sr, err := c.Search(ctx, pnCollection, []string{}, idFilter, []string{"ID"}, []entity.Vector{vectorParam}, "Vector",
		entity.IP, 500, sp)
```
i want first filter by the ids(may be more than 5k),  then find the best 500 result by the vectorParam using `entity.IP,`

3. i have tried  it myself with milvus-2.0  , it can give the wanted result , but the request will use more than 1000ms , this can not work online, i want to know is there someway to improve it .

4. farther way , i want to know :
     1. is the ids of vector have been indexed ?
     2. the request will first do filter by ids then find the topK, is it?

sorry for bother you , really thank you for your time!

**Comment by xiaofan-luan** [comment]
> sorry for bother you
> 
> ## I want use milvus help me do this work
> 1. i have a collection have about 6million vectors like this
>    ![image](https://user-images.githubusercontent.com/13183268/141223846-4fde1909-120f-4a6c-95f1-75b9d0067a57.png)
> 2. my query is like this
> 
> ```
>         vectorParam := entity.FloatVector(some_test_vector)  // 64 dimensions  
> 	idFilter := "ID in [ 123,234,456 ]"  // the ids may be more than 5000
> 	// Use flat search param
> 	sp, _ := entity.NewIndexFlatSearchParam(10)
> 	sr, err := c.Search(ctx, pnCollection, []string{}, idFilter, []string{"ID"}, []entity.Vector{vectorParam}, "Vector",
> 		entity.IP, 500, sp)
> ```
> 
> i want first filter by the ids(may be more than 5k), then find the best 500 result by the vectorParam using `entity.IP,`
> 
> 1. i have tried  it myself with milvus-2.0  , it can give the wanted result , but the request will use more than 1000ms , this can not work online, i want to know is there someway to improve it .
> 2. farther way , i want to know :
>    
>    1. is the ids of vector have been indexed ?
>    2. the request will first do filter by ids then find the topK, is it?
> 
> sorry for bother you , really thank you for your time!

Hi chen, one thing you can try is to  collection.query to find all the original vectors. they try to use utility.calc_distance to calculate distance of the these vectors with your original vectors and sort by your self.

Since we don't have kv index so far, maybe the bottleneck here is to get vectors from disk, you will need to verify if the change can match your requirement.

A better solution is to change the execution path inside search , this is already on our roadmap, the rule based optimizer can help improve the hybrid search performance， that will probably be done by the end of the year (milvus 2.1)

**Reply by chenchuangc** [reply]
really really thank you so much for your quick response!!!

i have just try the 
```
        vector :=  []float32{xxxx}  // 64 dim
	intIds := []int64{12,  23,  456,  xxx}  // may be 5000 ids
	r, err := c.CalcDistance(ctx, pnCollection, []string{}, entity.IP, entity.NewColumnInt64("Vector", intIds), entity.NewColumnFloatVector("Vector", 64, [][]float32{vector}))

```
5000 ids and a 64 dimensions vector , the request time is more than 500ms ,  it is still  too long for me (i want the 90 percent request time is below 80ms).  

Is there some settings that can store all the vectors in memory( i have enough memory ) , my test milvus  cluster is single-node way .

> A better solution is to change the execution path inside search , this is already on our roadmap,

thank you for your advice , but my recall stage is multi-way, milvus can not satisfied my requirements.

**Reply by xiaofan-luan** [reply]
have you ever tried to query by id of 5000 ids and see how long does it take?

**Comment by chenchuangc** [comment]
i think that in the singl-node way , if the collection is loaded, the vectors is already in memory , am i right?

![image](https://user-images.githubusercontent.com/13183268/141250343-e7f15bfa-da83-4eca-8d0a-918e8d501348.png)

@xiaofan-luan

**Reply by xiaofan-luan** [reply]
based on what index you used. If IVF FLAT yes, most of the vector index, for example IVF PQ only load index
